src/main/php/web/session/CookieBased.class.php

```diff
@@ -0,0 +1,165 @@
+<?php namespace web\session;
+
+use lang\FormatException;
+use util\Secret;
+use web\Cookie;
+use web\session\cookie\{Session, Format};
+
+class CookieBased extends Sessions {
+  private $key, $format;
+  private $attributes= [
+    'path'     => '/',
+    'secure'   => true,
+    'domain'   => null,
+    'httpOnly' => true,
+    'sameSite' => 'Lax'
+  ];
+
+  /**
+   * Creates an new cookie-based session
+   *
+   * @param  string|util.Secret $key
+   * @param  ?web.session.cookie.Format $format
+   */
+  public function __construct($key, Format $format= null) {
+    $this->key= $key instanceof Secret ? $key : new Secret($key);
+    $this->format= $format ?? Format::available()[0];
+  }
+
+  /**
+   * Sets the session transport
+   *
+   * @param  web.session.Transport $transport
+   * @return self
+   */
+  public function via($transport) {
+
+    // Special-case handling for Cookies-transport
+    if ($transport instanceof Cookies) {
+      $this->attributes= $transport->attributes();
+    }
+
+    return parent::via($transport);
+  }
+
+  /**
+   * Switch whether to also transmit via insecure connections (HTTP).
+   *
+   * @param  bool $insecure
+   * @return self
+   */
+  public function insecure($insecure= true) {
+    $this->attributes['secure']= !$insecure;
+    return $this;
+  }
+
+  /**
+   * Locates an existing and valid session; returns NULL if there is no such session.
+   *
+   * @param  web.Request $request
+   * @return ?web.session.ISession
+   */
+  public function locate($request) {
+    $cookie= $request->cookie($this->name);
+    return null === $cookie ? null : $this->open($cookie);
+  }
+
+  /**
+   * Attaches session to response 
+   *
+   * @param  web.session.ISession $session
+   * @param  web.Response $response
+   * @return void
+   */
+  public function attach($session, $response) {
+    $response->cookie((new Cookie($this->name, $session->id()))
+      ->maxAge($this->duration())
+      ->path($this->attributes['path'])
+      ->secure($this->attributes['secure'])
+      ->domain($this->attributes['domain'])
+      ->httpOnly($this->attributes['httpOnly'])
+      ->sameSite($this->attributes['sameSite'])
+    );
+  }
+
+  /**
+   * Detaches session from response 
+   *
+   * @param  web.session.ISession $session
+   * @param  web.Response $response
+   * @return void
+   */
+  public function detach($session, $response) {
+    $response->cookie((new Cookie($this->name(), null))
+      ->path($this->attributes['path'])
+      ->domain($this->attributes['domain'])
+    );
+  }
+
+  /**
+   * URL-safe base64 encoding
```

Comment:
Base64 adds some overhead. As an example, a 1664 bytes JSON data structure is encoded as 2220 bytes.

Reply:
gzcompress()'ing the data before encoding it with base64 would yield only 856 bytes.

Reply:
...but would add another dependency.

Reply:
Huffman coding is relatively easy to implement and gives moderate savings without requiring an extra PHP extension compiled in:

* JSON value: 2816 bytes
* Encrypted and encoded: 3809 bytes
* If we compress it before: 2102 bytes, then after encrypting and encoding: 2857 bytes (*almost a kilobyte saved, 75% of the size*)